### DISCUSSION TITLE
Why do we both need `fromMarkdownExtensions` and `micromarkExtensions`?

### DISCUSSION THREAD

**Discussion by crutchcorn** [remarkjs/remark#869] [Q&A] [answered]
I don't understand why we need both `micromark` and `mdast` extensions in many remark plugins.

Take `remark-gfm`, for example:

https://github.com/remarkjs/remark-gfm/blob/main/index.js#L17-L19

```javascript
  add('micromarkExtensions', gfm(options))
  add('fromMarkdownExtensions', gfmFromMarkdown())
  add('toMarkdownExtensions', gfmToMarkdown(options))
```

As I understand it (and I'm certain I'm wrong - looking for guidance), you would do something like this in Remark:

```
import {readSync} from 'to-vfile'
import {reporter} from 'vfile-reporter'
import {unified} from 'unified'
import remarkParse from 'remark-parse'
import remarkFrontmatter from 'remark-frontmatter'
import remarkStringify from 'remark-stringify'

const file = readSync('example.md')

unified()
  .use(remarkParse)
  .use(remarkStringify)
  .use(remarkFrontmatter, ['yaml', 'toml'])
  .use(() => (tree) => {
    console.dir(tree)
  })
  .process(file)
  .then((file) => {
    console.error(reporter(file))
    console.log(String(file))
  })
```

This would start a few things:

1) [`remarkParse` would call `mdast-util-from-markdown` under-the-hood to turn it into `mdast` format](https://github.com/remarkjs/remark/blob/main/packages/remark-parse/lib/index.js#L6)
2) `remarkFrontmatter` then changes this AST in some way (??? this is the part I'm unclear about)
3) [`remark-stringify` finally calls `mdast-util-to-markdown` to render back into markdown](https://github.com/remarkjs/remark/blob/main/packages/remark-stringify/lib/index.js)

I'm not entirely sure about `2` and feel there's some required understanding of how `micromark` comes into play with all of this that I'd need to learn more about before fully grasping the code that's happening in step 2.

Does [`mdast-util-from-markdown`](https://github.com/syntax-tree/mdast-util-from-markdown) and [`micromark`](https://github.com/micromark/micromark) not have some level of overlap, as they both (at first glance) appear to be markdown parsers? Why do we need to have both extensions registered in a remark plugin? What is `micromark` doing - since it appears that Remark is using `mdast` utils under-the-hood?

**Comment by ChristianMurphy** [comment]
Some clarification on the parsing:
* micromark parses tokens, a list of characters and their token types.
* remark takes the tokens and constructs an abstract syntax tree

And on the compiling:
* micromark can compile tokens markdown tokens to HTML
* remark can compile an AST to HTML or back to markdown

And on parser/compiler plugins, like frontmatter:
* parser plugins need both a tokenizer to categorize what each character does, and an AST parser to take the tokens to a tree format
* most syntax plugins also include code to compile the new tokens and tree nodes back to markdown as well

For more see the micromark architecture doc https://github.com/micromark/micromark#architecture the guide to syntax plugins  
https://github.com/micromark/micromark#extending-markdown and the unified guides http://unifiedjs.com/learn/

**Comment by wooorm** [comment]
Forgetting remark for a second, and just looking at micromark and mdast, they do different things:

- micromark itself is explained as Christian mentions in [§ Architecture](https://github.com/micromark/micromark#architecture). The last part, compiling to HTML, is something we skip in this case because we’re doing ASTs.
- with mdast, we do take micromarks first steps, but instead turn the “events” into an AST. This is what `mdast-util-from-markdown` does. Then there’s a new step, where the AST can be changed. And at the end we take the AST and turn it into markdown (`mdast-util-to-markdown`). ASTs are much easier to work with compared to tokens and events.

There are several places to extend this process:

- at micromark, when parsing, as a [syntax extension](https://github.com/micromark/micromark#architecture) — this takes care of taking nonstandard markdown and turning it into events (e.g., GFM, frontmatter, etc)
- at micromark, when serializing, as an [html extension](https://github.com/micromark/micromark#htmlextension) — these takes care of turning events into HTML, e.g., turning GFM table events into a `<table>`
- mdast, when parsing (`fromMarkdown`), to take events and turn them into a syntax tree
- mdast, when serializing (`toMarkdown`), to take a syntax tree and turn it into markdown

Sometimes it’s useful to not have to deal with syntax trees, but compile straight to HTML. That’s when micromark can be used on its own.
Sometimes its useful to manually parse some ASTs, that’s where {from,to}-markdown come into play, but from-markdown can defer much work to micromark.
Sometimes it’s useful to just deal with ASTs and don’t have to worry about parsing. That’s where remark/unified/etc are. They abstract away all that underlying stuff.

---

> Does mdast-util-from-markdown and micromark not have some level of overlap, as they both (at first glance) appear to be markdown parsers? Why do we need to have both extensions registered in a remark plugin? What is micromark doing - since it appears that Remark is using mdast utils under-the-hood?
 
They have a lot of overlap! But not a complete overlap. So extensions are needed for all 3

**Comment by about-code** [comment, marked as answer]
Sorry for adding a late answer but I hope a bit of an of ASCII Art ([CC BY-SA](https://creativecommons.org/licenses/by-sa/3.0/#)) might help in understanding where these bits and pieces fit in the processing of markdown files.

~~~
1. Input            | .~~~~ ['#','F','o','o','EOL'] <~~~ file.md
                    | |
2. Tokenize         | `~~~~ micromark(string)~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
                    |                                                          o)-- micromarkExtension()
3. Token Stream     | .~~ ['enter:atxHeading','F','o','o','exit:atxHeading'] ~~'
                    | |
4. Grow Syntax Tree | `~~~> mdast-util-from-markdown(tokens) ~~~~~~~~~~~~~~~~~~.
                    |                                                 o        o)-- fromMarkdownExtension()
5. Transform Tree   | .~~~~~ transform(treeNode) => treeNode ~~~~~~~ / \  ~~~~~'
                    | |                                             o   o
6. Serialize        | `~~~~ mdast-util-to-markdown(tree) ~~~~~~~~~~~~~~~~~~~~~~.
                    |                                                          o)-- toMarkdownExtension()
7. Output           | <~~~~ ['#','F','o','o','EOL'] <~~~~~~~~~~~~~~~~~~~~~~~~~~'
~~~
